### DISCUSSION TITLE
Inserting YAML front matter

### DISCUSSION THREAD

**Discussion by djole21** [vsch/flexmark-java#568] [Q&A]
What would be the best way to insert yaml front matter into markdown document if it does not exist?

**Comment by marad** [comment]
@djole21 did you find the way to do this? Can you share?

**Reply by djole21** [reply]
It was something like this. I hope it helps.


    public String addOrUpdateFrontMatter(String markdownContent) {
        String frontMatterDelimiter = "---";
        Map<String, Object> data = new LinkedHashMap<>();

        Map<String, List<String>> frontMatter = getYamlFrontDataMatter(markdownContent);

        Yaml yaml = createYamlParser();
        Matcher matcher = getMatcher(markdownContent);
        String body;
        boolean matcherFound = matcher.find();
        // add edge case, document can have empty yaml front matter,
        // or it can be located somewhere in document but not at the beginning
        if (matcherFound && !frontMatter.isEmpty() || matcherFound && markdownContent.startsWith("---")) {
            String yamlFrontMatter = matcher.group(1);
            body = matcher.replaceFirst("").trim();
            data = yaml.load(yamlFrontMatter);
            // yaml.load() returns null if there are no items
            data = data == null ? new LinkedHashMap<>() : data;
            data.merge("header", "headerValue", (old, newValue) -> newValue);
        } else {
            data.put("header", "headerValue");
            body = markdownContent;
        }

        StringWriter writer = new StringWriter();
        yaml.dump(data, writer);

        return frontMatterDelimiter + "\n" + writer.toString().trim() + "\n" + frontMatterDelimiter + "\n" + body;
    }
    private Map<String, List<String>> getYamlFrontDataMatter(String markdownContent) {
        MutableDataSet options = new MutableDataSet();
        options.set(Parser.EXTENSIONS, List.of(YamlFrontMatterExtension.create()));
        Parser parser = Parser.builder(options).build();

        Node document = parser.parse(markdownContent);

        AbstractYamlFrontMatterVisitor visitor = new AbstractYamlFrontMatterVisitor();
        visitor.visit(document);
        return visitor.getData();
    }

    private static Yaml createYamlParser() {
        DumperOptions dumperOptions = new DumperOptions();
        //        dumperOptions.setIndent(2);
        dumperOptions.setPrettyFlow(true);
        dumperOptions.setDefaultFlowStyle(DumperOptions.FlowStyle.BLOCK);
        return new Yaml(dumperOptions);
    }

    private Matcher getMatcher(String markdownContent) {
        Pattern pattern = Pattern.compile("^---(.*)^---", Pattern.DOTALL | Pattern.MULTILINE);
        return pattern.matcher(markdownContent);
    }
